tests/clients/test_fenda_ditto_client.py

```diff
@@ -0,0 +1,171 @@
+    fenda_ditto_client.lam = 1.0
+
+    params = [val.cpu().numpy() for _, val in fenda_ditto_client.global_model.state_dict().items()]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(4)
+
+    # Training: both FENDA models are updated
+    state_dict = fenda_ditto_client.global_model.base_module.state_dict()
+    for key in state_dict:
+        state_dict[key] += 0.1
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(state_dict)
+    fenda_ditto_client.model.second_feature_extractor.load_state_dict(state_dict)
+
+    preds = {"global": torch.tensor([[1.0, 0.0], [0.0, 1.0]]), "local": torch.tensor([[1.0, 0.0], [0.0, 1.0]])}
+    target = torch.tensor([[1.0, 0.0], [1.0, 0.0]])
+    training_loss = fenda_ditto_client.compute_training_loss(preds, {}, target)
+    fenda_ditto_client.global_model.eval()
+    fenda_ditto_client.model.eval()
+    evaluation_loss = fenda_ditto_client.compute_evaluation_loss(preds, {}, target)
+    print(evaluation_loss)
+    assert isinstance(training_loss.backward, dict)
+    assert pytest.approx(13.673, abs=0.01) == training_loss.backward["backward"].item()
+    assert pytest.approx(0.8132616, abs=0.0001) == evaluation_loss.checkpoint.item()
+    assert evaluation_loss.checkpoint.item() != training_loss.backward["backward"].item()
+
+
+@pytest.mark.parametrize("type,model", [(FendaDittoClient, FendaModel(FeatureCnn(), FeatureCnn(), FendaHeadCnn()))])
+def test_compute_loss_freeze_global_feature_extractor(get_client: FendaDittoClient) -> None:  # noqa
+    torch.manual_seed(42)
+    fenda_ditto_client = get_client
+    fenda_ditto_client.global_model = SequentiallySplitExchangeBaseModel(FeatureCnn(), HeadCnn())
+    fenda_ditto_client.parameter_exchanger = FullParameterExchanger()
+    config: Config = {"current_server_round": 2}
+    fenda_ditto_client.criterion = torch.nn.CrossEntropyLoss()
+    fenda_ditto_client.lam = 1.0
+    fenda_ditto_client.freeze_global_feature_extractor = True
+
+    params = [val.cpu().numpy() for _, val in fenda_ditto_client.global_model.state_dict().items()]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(4)
+
+    # Training: only local FENDA model is updated
+    state_dict = fenda_ditto_client.global_model.base_module.state_dict()
+    for key in state_dict:
+        state_dict[key] += 0.1
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(state_dict)
+
+    preds = {"global": torch.tensor([[1.0, 0.0], [0.0, 1.0]]), "local": torch.tensor([[1.0, 0.0], [0.0, 1.0]])}
+    target = torch.tensor([[1.0, 0.0], [1.0, 0.0]])
+    training_loss = fenda_ditto_client.compute_training_loss(preds, {}, target)
+    fenda_ditto_client.global_model.eval()
+    fenda_ditto_client.model.eval()
+    evaluation_loss = fenda_ditto_client.compute_evaluation_loss(preds, {}, target)
+    print(evaluation_loss)
```

Comment:
Dropping print statements in tests is good practice 🙂

Reply:
removed